File: LeetCode/1458/golang/lc1458.go
package lc1458

import "math"

// Runtime complexity: O(L1*L2)
// Auxiliary space: O(L1*L2)
// Subjective level: medium+
// Solved on: 2026-01-08
func maxDotProduct(nums1 []int, nums2 []int) int {
	Y := len(nums1)
	X := len(nums2)

	// dp[y][x] holds max dot product of all possible subsequences from: nums1[0:y), nums2[0:x).
	// Initially set to minimum possible integer value.
	dp := make([][]int, Y+1)
	for y := range Y + 1 {
		dp[y] = make([]int, X+1)
		for x := range X + 1 {
			dp[y][x] = math.MinInt32
		}
	}

	// Calculate the actual max dot product of all possible subsequences; incrementally:
	// row-by-row, column-by-column in each row.
	for y := range Y {
		for x := range X {
			dp[y+1][x+1] = max(dp[y+1][x], dp[y][x+1], nums1[y]*nums2[x]+max(0, dp[y][x]))
		}
	}
	return dp[Y][X]
}

File: LeetCode/1458/cpp/lc1458.cpp
#include <algorithm>
#include <climits>
#include <cassert>
#include <vector>

class Solution {
public:
  int maxDotProduct(std::vector<int>& nums1, std::vector<int>& nums2) {

    // dp[x][y] stores max dot product of subsequences nums1[0...x], nums2[0...y].
    std::vector<std::vector<int>> dp(nums1.size() + 1, std::vector<int>(nums2.size() + 1, INT_MIN));
    for (int i = 0; i < nums1.size(); i++) {
      for (int j = 0; j < nums2.size(); j++) {
        dp[i+1][j+1] = std::max({dp[i][j+1], dp[i+1][j], std::max(0, dp[i][j]) + nums1[i] * nums2[j]});
      }
    }
    return dp[nums1.size()][nums2.size()];
  }
};

int main() {
  Solution s;
  std::vector<int> example1_1 = {2, 1, -2, 5};
  std::vector<int> example1_2 = {3, 0, -6};
  assert(s.maxDotProduct(example1_1, example1_2) == 18); // [2, -2] dot [3, -6]

  std::vector<int> example2_1 = {3, -2};
  std::vector<int> example2_2 = {2, -6, 7};
  assert(s.maxDotProduct(example2_1, example2_2) == 21); // [3] dot [7]

  std::vector<int> example3_1 = {-1, -1};
  std::vector<int> example3_2 = {1, 1};
  assert(s.maxDotProduct(example3_1, example3_2) == -1); // [-1] dot [1]
}
